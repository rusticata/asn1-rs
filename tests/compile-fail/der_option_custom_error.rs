use asn1_rs::{Error, FromDer, ParseResult};

pub struct CustomError;

impl From<Error> for CustomError {
    fn from(_value: Error) -> Self {
        CustomError
    }
}

pub type T = Option<u32>;

// NOTE: this fails, because Option::FromDer is only available with default
// error type.
// It cannot be implemented for custom errors, because it would conflict with
// default implementation because of default infaillible T -> Option<T> implementation
fn should_fail_for_now<'a>(input: &'a [u8]) -> ParseResult<'a, T, CustomError> {
    T::from_der(input)
}

fn main() {
    let _t = should_fail_for_now(&[0x00, 0x01, 0x02, 0x03]);
}
