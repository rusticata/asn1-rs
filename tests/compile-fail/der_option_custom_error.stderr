error[E0308]: mismatched types
  --> tests/compile-fail/der_option_custom_error.rs:18:5
   |
17 | fn should_fail_for_now<'a>(input: &'a [u8]) -> ParseResult<'a, T, CustomError> {
   |                                                ------------------------------- expected `Result<(&'a [u8], Option<u32>), asn1_rs::Err<CustomError>>` because of return type
18 |     T::from_der(input)
   |     ^^^^^^^^^^^^^^^^^^ expected `CustomError`, found `asn1_rs::Error`
   |
   = note: expected enum `Result<(&'a [u8], Option<u32>), asn1_rs::Err<CustomError, CustomError>>`
              found enum `Result<(&[u8], Option<u32>), asn1_rs::Err<asn1_rs::Error, asn1_rs::Error>>`
